piv_identity.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"crypto"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+
+	"golang.org/x/crypto/ssh/terminal"
+
+	"github.com/github/certstore"
+	"github.com/go-piv/piv-go/piv"
+	"github.com/pkg/errors"
+)
+
+// PivIdentities enumerates identities stored in the signature slot inside hardware keys
+func PivIdentities() ([]PivIdentity, error) {
+	cards, err := piv.Cards()
+	if err != nil {
+		return nil, err
+	}
+	var identities []PivIdentity
+	for _, card := range cards {
+		yk, err := piv.Open(card)
+		if err != nil {
+			continue
+		}
+		cert, err := yk.Certificate(piv.SlotSignature)
+		if err != nil {
+			continue
+		}
+		if cert != nil {
+			ident := PivIdentity{card: card, yk: yk}
+			identities = append(identities, ident)
+		}
+	}
+	return identities, nil
+}
+
+// PivIdentity is an entity identity stored in a hardware key PIV applet
+type PivIdentity struct {
+	card string
+	//pin  string
+	yk *piv.YubiKey
+}
+
+var _ certstore.Identity = (*PivIdentity)(nil)
+var _ crypto.Signer = (*PivIdentity)(nil)
+
+// Certificate implements the certstore.Identity interface
+func (ident *PivIdentity) Certificate() (*x509.Certificate, error) {
+	return ident.yk.Certificate(piv.SlotSignature)
+}
+
+// CertificateChain implements the certstore.Identity interface
+func (ident *PivIdentity) CertificateChain() ([]*x509.Certificate, error) {
+	return []*x509.Certificate{}, nil
+}
+
+// Signer implements the certstore.Identity interface
+func (ident *PivIdentity) Signer() (crypto.Signer, error) {
+	return ident, nil
+}
+
+// Delete implements the certstore.Identity interface
+func (ident *PivIdentity) Delete() error {
+	panic("deleting identities on PIV applet is not supported")
+}
+
+// Close implements the certstore.Identity interface
+func (ident *PivIdentity) Close() {
+	_ = ident.yk.Close()
+}
+
+// Public implements the crypto.Signer interface
+func (ident *PivIdentity) Public() crypto.PublicKey {
+	cert, err := ident.Certificate()
+	if err != nil {
+		return nil
+	}
+
+	return cert.PublicKey
+}
+
+// Sign implements the crypto.Signer interface
+func (ident *PivIdentity) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	private, err := ident.yk.PrivateKey(piv.SlotSignature, ident.Public(), piv.KeyAuth{
+		PINPrompt: ident.promptHardwareKeyPin,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get private key for signing")
+	}
+
+	switch private.(type) {
+	case *piv.ECDSAPrivateKey:
+		return private.(*piv.ECDSAPrivateKey).Sign(rand, digest, opts)
+	default:
+		return nil, fmt.Errorf("invalid key type")
+	}
+}
+
+func (ident *PivIdentity) promptHardwareKeyPin() (string, error) {
+	tty, err := os.OpenFile("/dev/tty", os.O_RDWR, fs.ModeCharDevice)
```

Comment:
What is the behavior of this if a TTY is not available?

For example: let's suppose someone is committing with a GUI like GitHub Desktop. There won't be a TTY where someone can type the device PIN. Will this error or will it hang waiting for someone to type a pin?

I'm less worried about supporting the GUI with a modal, I just want to understand the behavior.

Further, I suppose this will not work on Windows, correct? Windows' will import certificates in to its certificate store with the Certificate Propagation Service (or middleware). Should this functionality be omitted for Windows?

Reply:
I don't think it's going to work very well on other platforms or when using a GUI git client.
I realized I might've submitted this PR a little too early, since I just also found out that I'm not handling the certificate chain the right way either.

For context though, it seems like the PIN prompt is also a big issue with `gpgsm`.
Looking at the source code there, there's a whole library dedicated to handle the PIN prompt.

What I did here is extract the most basic and simplified scenario from that library so it'll at least be functional when using a shell.

Reply:
@vcsjones I did a little more research on how this procedure usually works, and decided to go with a different route to get the PIN when using a Yubikey.

If a PIN is required now, `smimesign` will try to get it via the `pinentry` program.
`pinentry` is part of GnuPG tools and is considered to be a more secure way of getting such information from the user.

It also has a lot of different implementations for different platforms, so the user can choose which program to use if they want to.

I've tested this PR using both the terminal and Github Desktop on a MacBook.
I didn't get a chance to test it on other platforms yet.